test/test_tx.py

```diff
@@ -0,0 +1,171 @@
+parser.add_argument('--framesperburst', dest="N_FRAMES_PER_BURST", default=1, type=int)
+parser.add_argument('--delay', dest="DELAY_BETWEEN_BURSTS", default=500, type=int,
+                    help="delay between bursts in ms")
+parser.add_argument('--mode', dest="FREEDV_MODE", type=str, choices=['datac0', 'datac1', 'datac3'])
+parser.add_argument('--audiodev', dest="AUDIO_OUTPUT_DEVICE", default=-1, type=int,
+                    help="audio output device number to use, use -2 to automatically select a loopback device") 
+parser.add_argument('--list', dest="LIST", action="store_true", help="list audio devices by number and exit")  
+parser.add_argument('--testframes', dest="TESTFRAMES", action="store_true", default=False, help="list audio devices by number and exit")  
+
+
+args = parser.parse_args()
+
+if args.LIST:
+    p = pyaudio.PyAudio()
+    for dev in range(0,p.get_device_count()):
+        print("audiodev: ", dev, p.get_device_info_by_index(dev)["name"])
+    quit()
+
+N_BURSTS = args.N_BURSTS
+N_FRAMES_PER_BURST = args.N_FRAMES_PER_BURST
+DELAY_BETWEEN_BURSTS = args.DELAY_BETWEEN_BURSTS/1000
+AUDIO_OUTPUT_DEVICE = args.AUDIO_OUTPUT_DEVICE
+
+MODE = codec2.FREEDV_MODE[args.FREEDV_MODE].value
+
+# AUDIO PARAMETERS
+AUDIO_FRAMES_PER_BUFFER = 2400
+MODEM_SAMPLE_RATE = codec2.api.FREEDV_FS_8000
+AUDIO_SAMPLE_RATE_TX = 48000
+assert (AUDIO_SAMPLE_RATE_TX % MODEM_SAMPLE_RATE) == 0
+
+# check if we want to use an audio device then do an pyaudio init
+if AUDIO_OUTPUT_DEVICE != -1: 
+    p = pyaudio.PyAudio()
+    # auto search for loopback devices
+    if AUDIO_OUTPUT_DEVICE == -2:
+        loopback_list = []
+        for dev in range(0,p.get_device_count()):
+            if 'Loopback: PCM' in p.get_device_info_by_index(dev)["name"]:
+                loopback_list.append(dev)
+        if len(loopback_list) >= 2:
+            AUDIO_OUTPUT_DEVICE = loopback_list[1] #0  = RX   1 = TX
+            print(f"loopback_list tx: {loopback_list}", file=sys.stderr)
+        else:
+            quit()        
+    print(f"AUDIO OUTPUT DEVICE: {AUDIO_OUTPUT_DEVICE} DEVICE: {p.get_device_info_by_index(AUDIO_OUTPUT_DEVICE)['name']} \
+            AUDIO SAMPLE RATE: {AUDIO_SAMPLE_RATE_TX}", file=sys.stderr)
+
+    # pyaudio init
+    stream_tx = p.open(format=pyaudio.paInt16,
+                            channels=1,
+                            rate=AUDIO_SAMPLE_RATE_TX,
+                            frames_per_buffer=AUDIO_FRAMES_PER_BUFFER, #n_nom_modem_samples
+                            output=True,
+                            output_device_index=AUDIO_OUTPUT_DEVICE
+                            )      
+                                
+                                
+resampler = codec2.resampler()
+
+# data binary string
+if args.TESTFRAMES:
+    data_out = bytearray(14)
+    data_out[:1]   = bytes([255])
+    data_out[1:2]  = bytes([1])
+    data_out[2:]  = b'HELLO WORLD'
+    
+else:
+    data_out = b'HELLO WORLD!'
```

Comment:
Option for using the test programs for checking the TNC itself. Just append `--testframes` as additional parameter. This could be used maybe as standard instead of sending a simple b'HELLO WORLD!'